addons/mail/models/mail_thread.py

```diff
@@ -2332,83 +2368,103 @@ def send_notifications():
+    def _notify_by_email_prepare_rendering_context(self, message, msg_vals, model_description=False,
+                                                   force_email_company=False, force_email_lang=False):
         """ Prepare rendering context for notification email.
 
         Signature: if asked a default signature is computed based on author. Either
         it has an user and we use the user's signature. Either we do not find any
         user and we compute a default one based on the author's name.
 
         Company: either there is one defined on the record (company_id field set
-        with a value), either we use env.company.
+        with a value), either we use env.company. A new parameter allows to force
+        its value.
 
         Lang: when calling this method, ``_fallback_lang`` should already been
         called, or a lang set in context with another way. A wild guess is done
         based on templates to try to retrieve the recipient's language when a flow
         like "send by email" is performed. Lang is used to try to have the
-        notification layout in the same language as the email content.
+        notification layout in the same language as the email content. A new
+        parameter allows to force its value.
 
         :param model_description: model description used in email notification process
           (computed if not given);
+        :param force_email_company: res.company record used when rendering notification
+          layout. Otherwise computed based on current record;
+        :param force_email_lang: lang used when rendering content, used notably to
+          compute model name;
         """
+        # compute send user and its related signature; try to use self.env.user instead of browsing
+        # user_ids if he is the author will give a sudo user, improving access performances and cache usage.
         signature = ''
-        user = self.env.user
-        author = message.env['res.partner'].browse(msg_vals.get('author_id')) if msg_vals else message.author_id
-        model = msg_vals.get('model') if msg_vals else message.model
-        add_sign = msg_vals.get('add_sign') if msg_vals else message.add_sign
-        subtype_id = msg_vals.get('subtype_id') if msg_vals else message.subtype_id.id
-        message_id = message.id
-        record_name = msg_vals.get('record_name') if msg_vals else message.record_name
-        author_user = user if user.partner_id == author else author.user_ids[0] if author and author.user_ids else False
-        # trying to use user (self.env.user) instead of browing user_ids if he is the author will give a sudo user,
-        # improving access performances and cache usage.
-        if author_user:
-            user = author_user
-            if add_sign:
-                signature = user.signature
-        else:
-            if add_sign:
+        add_sign = msg_vals.get('add_sign') if msg_vals and 'add_sign' in msg_vals else message.add_sign
```

Comment:
Seems like the code is doing a lot of effort to make "msg_vals" optional yet it's a required parameter?

Reply:
Could be msg_vals=None indeed.